File: bot/bonobot.py
#!/usr/bin/env python
import trio
from math import atan2, pi
from bot import Bot
from random import randint

WEB_URL = 'https://lemuria.7185.fr/api/v1'
WS_URL = 'wss://lemuria.7185.fr/api/v1/ws'

class Bonobot(Bot):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.name = '[bonobot]'
        self.following = None
        self.move_speed = 1
        self.current_move_thread = 0
        self.logging_enabled = False

    async def move(self, dest_x: float, dest_z: float) -> None:
        thread_id = self.current_move_thread
        tick = 200
        dx = dest_x - self.x
        dz = dest_z - self.z
        length = (dx ** 2 + dz ** 2) ** 0.5
        direction = atan2(dx, dz) + pi
        self.state = 'walk'
        num_step = int(length * (1 / self.move_speed))
        if num_step > 0:
            x_gap = dx / num_step
            z_gap = dz / num_step
            gaps = [[self.x + i * x_gap, self.z + i * z_gap] for i in range(1, num_step + 1)]
        else:
            gaps = [[dest_x, dest_z]]
        for step in gaps:
            if thread_id != self.current_move_thread:
                break
            self.set_position(step[0], self.y, step[1], yaw=direction)
            await self.send_position()
            await trio.sleep(tick / 1e3)
        self.state = 'idle'
        await self.send_position()

    async def on_connected(self) -> None:
        await self.change_avatar(self.avatar)
        await self.get_world_list()
        for i, w in self.worldlist.items():
            if w['name'] == 'Village':
                await self.world_enter(i)
        await self.send_msg('hello')
        await self.send_position()

    async def on_user_join(self, msg: str) -> None:
        print(f"* {msg} joined")

    async def on_user_part(self, msg:str) -> None:
        print(f"* {msg} left")

    async def on_msg(self, user: str, msg: str) -> None:
        print(f"<{user}> {msg}")

        # Return early if the message is from the bot itself
        if user == self.name:
            return

        # Parse the message and extract the command and arguments
        command, *args = msg.split()

        # Handle the different commands
        if command == "!list":
            l = ' '.join([f"{u.name}({u.avatar}:{i})" for i, u in self.userlist.items()])
            await self.send_msg(l)
        elif command == "!pos":
            await self.send_msg(f"{self.x},{self.y},{self.z}")
        elif command == "!come":
            u = next((u for u in self.userlist.values() if u.name == user), None)
            if u is None:
                await self.send_msg("Sorry, I don't know who you are.")
            elif u.world != self.world:
                await self.send_msg(f"Sorry, I'm on {self.worldlist[self.world]['name'] if self.world else 'Nowhere'}...")
            else:
                await self.send_msg("Coming...")
                self.current_move_thread += 1
                self.nursery.start_soon(self.move, u.x, u.z)
        elif command == "!whereami":
            u = next((u for u in self.userlist.values() if u.name == user), None)
            if u is None:
                await self.send_msg("Sorry, I don't know who you are.")
            else:
                await self.send_msg(f"{u.x},{u.y},{u.z}")
        elif command == "!speed":
            try:
                value = int(args[0])
            except (IndexError, ValueError):
                value = 1
            self.move_speed = value
            await self.send_msg(f"Running at {self.move_speed}")
        elif command == "!change":
            await self.change_avatar(randint(0, 16))

b = Bonobot(WEB_URL, WS_URL)
b.run()

File: bot/src/bonobot.ts
import {Bot} from './bot.ts'
import {setTimeout} from 'timers/promises'

const WEB_URL = 'https://lemuria.7185.fr/api/v1'
const WS_URL = 'wss://lemuria.7185.fr/api/v1/ws'

class Bonobot extends Bot {
  following: string | null
  moveSpeed: number
  currentMoveThread: number

  constructor(webUrl: string, wsUrl: string) {
    super(webUrl, wsUrl, false) // loggingEnabled set to false
    this.name = '[bonobot]'
    this.following = null
    this.moveSpeed = 1
    this.currentMoveThread = 0
  }

  async move(destX: number, destZ: number): Promise<void> {
    const threadId = this.currentMoveThread
    const tick = 200
    const dx = destX - this.x
    const dz = destZ - this.z
    const length = Math.sqrt(dx ** 2 + dz ** 2)
    const direction = Math.atan2(dx, dz) + Math.PI
    this.state = 'walk'
    const numStep = Math.floor(length * (1 / this.moveSpeed))
    let gaps: [number, number][]
    if (numStep > 0) {
      const xGap = dx / numStep
      const zGap = dz / numStep
      gaps = Array.from({length: numStep}, (_, i) => [
        this.x + (i + 1) * xGap,
        this.z + (i + 1) * zGap
      ])
    } else {
      gaps = [[destX, destZ]]
    }
    for (const [stepX, stepZ] of gaps) {
      if (threadId !== this.currentMoveThread) {
        break
      }
      this.setPosition(stepX, this.y, stepZ, 0, direction, 0)
      await this.sendPosition()
      await setTimeout(tick)
    }
    this.state = 'idle'
    await this.sendPosition()
  }

  async on_connected(): Promise<void> {
    await this.changeAvatar(this.avatar)
    await this.getWorldList()
    for (const [i, w] of Object.entries(this.worldlist)) {
      if (w.name === 'Village') {
        await this.worldEnter(parseInt(i))
        break
      }
    }
    await this.sendMsg('hello')
    await this.sendPosition()
  }

  async on_user_join(msg: string): Promise<void> {
    const user = this.userlist[msg]
    console.log(
      `[${new Date().toISOString()}] * ${user ? user.name : msg} joined`
    )
  }

  async on_user_part(msg: string): Promise<void> {
    const user = this.userlist[msg]
    console.log(
      `[${new Date().toISOString()}] * ${user ? user.name : msg} left`
    )
  }

  async on_msg(userId: string, msg: string): Promise<void> {
    const user = this.userlist[userId]
    console.log(
      `[${new Date().toISOString()}] <${user ? user.name : msg}> ${msg}`
    )

    if (user?.name === this.name) {
      return
    }

    const [command, ...args] = msg.split(' ')

    switch (command) {
      case '!list': {
        const l = Object.entries(this.userlist)
          .map(([i, u]) => `${u.name}(${u.avatar}:${i})`)
          .join(' ')
        await this.sendMsg(l)
        break
      }
      case '!pos':
        await this.sendMsg(`${this.x},${this.y},${this.z}`)
        break
      case '!come': {
        if (!user) {
          await this.sendMsg("Sorry, I don't know who you are.")
        } else if (user.world !== this.world) {
          const worldName = this.world
            ? this.worldlist[this.world]?.name
            : 'Nowhere'
          await this.sendMsg(`Sorry, I'm on ${worldName}...`)
        } else {
          await this.sendMsg('Coming...')
          this.currentMoveThread++
          this.move(user.x, user.z).catch(console.error)
        }
        break
      }
      case '!whereami': {
        if (!user) {
          await this.sendMsg("Sorry, I don't know who you are.")
        } else {
          await this.sendMsg(`${user.x},${user.y},${user.z}`)
        }
        break
      }
      case '!speed': {
        this.moveSpeed = parseInt(args[0]) || 1
        await this.sendMsg(`Running at ${this.moveSpeed}`)
        break
      }
      case '!change':
        await this.changeAvatar(Math.floor(Math.random() * 17))
        break
    }
  }
}

const b = new Bonobot(WEB_URL, WS_URL)
b.run()
